src/OrbitBase/ThreadPoolTest.cpp

```diff
@@ -378,3 +380,120 @@ TEST(ThreadPool, WaitWithoutShutdown) {
       },
       "");
 }
+
+TEST(ThreadPool, FutureBasic) {
+  constexpr size_t kThreadPoolMinSize = 1;
+  constexpr size_t kThreadPoolMaxSize = 2;
+  constexpr absl::Duration kThreadTtl = absl::Milliseconds(5);
+  std::unique_ptr<ThreadPool> thread_pool =
+      ThreadPool::Create(kThreadPoolMinSize, kThreadPoolMaxSize, kThreadTtl);
+
+  absl::Mutex mutex;
+  bool called = false;
+
+  {
+    absl::MutexLock lock(&mutex);
+    orbit_base::Future<void> future = thread_pool->Schedule([&]() {
+      absl::MutexLock lock(&mutex);
+      called = true;
+    });
+
+    EXPECT_TRUE(future.IsValid());
+    EXPECT_FALSE(future.IsFinished());
+
+    EXPECT_TRUE(mutex.AwaitWithTimeout(absl::Condition(
+                                           +[](bool* called) { return *called; }, &called),
+                                       absl::Milliseconds(100)));
```

Comment:
Isn't it possible for the lambda to start executing before we release the mutex here, in which case we would lock twice? What is preventing the future from being finished by the time "EXPECT_FALSE(future.IsFinished());" is called?

Reply:
Ah nevermind, got it, the mutex inside the lambda will wait until we release it with AwaitWithTimeout